### PR TITLE
Update mimir-prometheus to `16653689899c` (2023-09-01)

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
This is a PR with the changes from https://github.com/grafana/mimir-prometheus/pull/526. A GEM image based on these changes has been running a dev cluster for a week now without issues.

These are the diffs
* https://github.com/grafana/mimir-prometheus/compare/002ae0aa1d7e...77ac7ad40
* https://github.com/prometheus/prometheus/compare/76dd9b547...b6f903b5f

There are a lot of changes in prometheus, some notable changes are
* https://github.com/grafana/mimir-prometheus/commit/3c80963e8102b9baaffb10dd76db9c9bfcd1c077 which mmaps head chunks in the background insteaed of on the append path
* https://github.com/grafana/mimir-prometheus/commit/8ef7dfdeebf0a7491973303c7fb6b68ec5cc065b which sets a 1KiB size limit on all chunks (in addition to the 120-sample limit)
* https://github.com/grafana/mimir-prometheus/commit/54aaa2bd7e9dad6a0eae34232b95b82dc7143059 which adds two new promQL functions for native histograms - `histogram_stdvar` and `histogram_stddev`

### CODE REVIEW

File: pkg/mimirpb/compat_test.go [hunk 1]
```diff
@@ -463,7 +463,7 @@ func TestFromHistogramToHistogramProto(t *testing.T) {
 	p := FromHistogramToHistogramProto(ts, h)
 
 	expected := Histogram{
-		Count:          &Histogram_CountInt{18},
```
Comment (dimitarvdimitrov) [inline pkg/mimirpb/compat_test.go:466]: this was necessary due to this change https://github.com/grafana/mimir-prometheus/commit/aa82fe198f66f4e2a41647fd0845834acf365e0e#diff-4c7063fbe57f049434ef7ed4d6360beb38a6cb4bd8154b2d5a2143557a5f52e3

File: pkg/mimirpb/compat_test.go [hunk 2]
```diff
@@ -511,7 +511,7 @@ func TestFromFloatHistogramToHistogramProto(t *testing.T) {
 	p := FromFloatHistogramToHistogramProto(ts, h)
 
 	expected := Histogram{
-		Count:          &Histogram_CountFloat{18},
```
Comment (dimitarvdimitrov) [inline pkg/mimirpb/compat_test.go:514]: this was necessary due to this change https://github.com/grafana/mimir-prometheus/commit/aa82fe198f66f4e2a41647fd0845834acf365e0e#diff-4c7063fbe57f049434ef7ed4d6360beb38a6cb4bd8154b2d5a2143557a5f52e3

File: pkg/util/test/histogram.go [hunk 3]
```diff
@@ -36,7 +36,7 @@ func GenerateTestGaugeFloatHistogram(i int) *histogram.FloatHistogram {
 // explicit decoded version of GenerateTestHistogram and GenerateTestFloatHistogram
 func GenerateTestSampleHistogram(i int) *model.SampleHistogram {
 	return &model.SampleHistogram{
-		Count: model.FloatString(10 + i*8),
+		Count: model.FloatString(12 + i*9),
```
Comment (dimitarvdimitrov) [inline pkg/util/test/histogram.go:39]: this was necessary due to this change https://github.com/grafana/mimir-prometheus/commit/aa82fe198f66f4e2a41647fd0845834acf365e0e#diff-4c7063fbe57f049434ef7ed4d6360beb38a6cb4bd8154b2d5a2143557a5f52e3

**Review by dimitarvdimitrov — COMMENTED** [review]
there were some changes necessary because test setup code in prometheus also changed

File: pkg/storage/tsdb/block/block_test.go [hunk 4]
```diff
@@ -206,7 +213,7 @@ func TestUpload(t *testing.T) {
 		require.Equal(t, 3, len(bkt.Objects()))
 		chunkFileSize := getFileSize(t, filepath.Join(tmpDir, b1.String(), ChunksDirname, "000001"))
 		require.Equal(t, chunkFileSize, int64(len(bkt.Objects()[path.Join(b1.String(), ChunksDirname, "000001")])))
-		require.Equal(t, 401, len(bkt.Objects()[path.Join(b1.String(), IndexFilename)]))
+		require.Equal(t, 418, len(bkt.Objects()[path.Join(b1.String(), IndexFilename)]))
```
Comment (dimitarvdimitrov) [inline pkg/storage/tsdb/block/block_test.go:216]: I wonder if instead we should instead just compare it with the size of the file that was written to disk. Asserting on the number of bytes in an index when testing bucket upload functionality seems flaky

File: vendor/github.com/prometheus/prometheus/tsdb/head.go [hunk 5]
```diff
@@ -494,6 +494,10 @@ func newHeadMetrics(h *Head, r prometheus.Registerer) *headMetrics {
 				60 * 60 * 12, // 12h
 			},
 		}),
+		mmapChunksTotal: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_tsdb_mmap_chunks_total",
```
Comment (pstibrany) [inline vendor/github.com/prometheus/prometheus/tsdb/head.go:498]: Do we want to expose this value in ingesters metrics and aggregate over all tenants?

**Review by pstibrany — APPROVED** [review]
Mimir changes look fine. I can't say I fully understand all incoming Prometheus changes though.

File: vendor/github.com/prometheus/prometheus/tsdb/head.go (hunk 5 above)
Comment (dimitarvdimitrov) [reply vendor/github.com/prometheus/prometheus/tsdb/head.go:498]: yeah, why not; updated in https://github.com/grafana/mimir/pull/5927/commits/314bb7279d1329dd1aed3c68a65889a54e236161

**Comment by dimitarvdimitrov** [comment]
I merged the PR in mimir-prometheus and updated the vendored commit SHA (which brought no additional code changes). This is ready for review/merge now

**Pull request merged**